gold/Gold_data.hpp

```diff
@@ -0,0 +1,271 @@
+
+#pragma once
+
+#include <cassert>
+#include <cstdint>
+
+#include <vector>
+#include <utility>
+#include <iostream>
+#include <algorithm>
+#include <functional>
+
+#include "lrand.hpp"
+
+class Gold_data {
+public:
+  void clear() {
+    device = false;
+    chunks.clear();
+  }
+
+  void set_addr(const uint64_t addr, const uint8_t sz) {
+    assert(!has_partial_overlap(addr, sz)); // No overlapping valid ranges
+    chunks.emplace_back(Chunk(addr, sz));
+
+    sort_chunks();
+  }
+
+  void add_addr(const uint64_t addr) {
+
+    for(auto i=0u;i<chunks.size();++i) {
+      auto &c = chunks[i];
+
+      if (c.is_hit(addr))
+        return; // done
+
+      if (c.addr-1 == addr) {
+        --c.addr;
+        assert(i>0 && !chunks[i-1].is_hit(addr-1));
+        c.data.insert(c.data.begin(), 1, rand_data.any());
+        return;
+      }
+
+      if ((c.addr+c.data.size())==addr) {
+        c.data.emplace_back(rand_data.any());
+
+        if (chunks.size()<i && chunks[i+1].addr == addr) { //merge with next
+          c.data.insert(c.data.end(), chunks[i+1].data.begin(), chunks[i+1].data.end());
+          ++i;
+          chunks.erase(chunks.begin()+i);
+        }
+        return;
+      }
+    }
+
```

Comment:
Doesn't seem to worry about what happens when chunks grows into each other.  Looks like you'll end up with inconsistent state (bugs).